### PR TITLE
Rewrite `Leaderboard` component to bring up to current code standards

### PR DESCRIPTION
**Pull request by peppy**
This started off as a fix for a bug I came across during manual testing, where the realm subscription wasn't actually subscribing via a certain flow. It turned in to a pretty heavy refactor/rewrite to remove unnecessary complexities. Seems valid effort though, given how central this component is to the game (and given that it will likely see new features and usages going forward).

To reproduce the failure I was witnessing, these steps should work:

- Set global leaderboard
- Switch to a beatmap with local scores
- Set local leaderboard
- Delete a local score
- Note that the leaderboard doesn't update

Looking at the original code, it should be quite clear how this can happen.

To fix this, I've moved the realm flow to use a single point of query. Unfortunately I'm not able to use the `ChangeSet` logic as it stands, due to the mod filtering that we apply to the realm result set (meaning that the indices of realm removals won't match up with our local scores list, see https://github.com/realm/realm-dotnet/issues/2752). This could potentially be worked around with a solution similar to `BeatmapCarousel` with two subscriptions, but as this is going to add some complexity I've not done that in this PR.

There's likely going to be a lot of areas pointed out that this can still be improved. I've got some future work I still want to do on this but I feel like I could spend many more hours on this before I would run out of improvements. Still, open to any suggestions that are seen as required.

Also note that the tests I added somewhat duplicate those in `TestSceneBeatmapLeaderboard`, but I intentionally wanted to cover the full flow, including affirming that the local play is recorded to the database, shows up at song select on returning, and can be interacted with. Nice high level coverage to have.

### CODE REVIEW

File: osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapLeaderboard.cs [hunk 1]
```diff
@@ -422,6 +416,8 @@ public void SetRetrievalState(PlaceholderState state)
             {
                 PlaceholderState = state;
             }
+
+            public void SetScores(ICollection<ScoreInfo> scores) => Scores = scores;
```
Comment (bdach) [inline osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapLeaderboard.cs:420]: I know it's test code but it's a bit of a weird one, can this one just be simple delegation to base? i.e.

```csharp
            public new ICollection<ScoreInfo> Scores
            {
                get => base.Scores;
                set => base.Scores = value;
            }
```

**Review by bdach — COMMENTED** [review]
Difficult diff to go through. I see an improvement in quality if I squint but it feels somewhat marginal if I'm honest - maybe because the diff is that large and not much was really changed, a lot of stuff has just been shuffled around.

Haven't ran yet, I have a few questions about the current code first.

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 2]
```diff
@@ -231,96 +145,183 @@ protected Leaderboard()
...
+        {
+            get => placeholderState;
+            set
             {
-                PlaceholderState = PlaceholderState.Retrieving;
-                loading.Show();
+                if (value == placeholderState)
+                    return;
+
+                loading.Hide();
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:244]: Yeah, this along with the mention you make of the `Reset` call shows how haphazard this still is. My end goal would be refactor this further to be a return state, or an alternative to calling `Scores=null` for fail cases.

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 3]
```diff
@@ -231,96 +145,183 @@ protected Leaderboard()
...
+                ChildrenEnumerable = scores.Select((s, index) => CreateDrawableScore(s, index + 1))
+            }, newFlow =>
+            {
+                scrollContainer.Add(scoreFlowContainer = newFlow);
+
+                int i = 0;
+
+                foreach (var s in scoreFlowContainer.Children)
+                {
+                    using (s.BeginDelayedSequence(i++ * 50))
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:317]: Would you prefer `i += 50` in this case?

File: osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapLeaderboard.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapLeaderboard.cs:420]: sure?

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 4]
```diff
@@ -126,62 +95,7 @@ public TScope Scope
...
-        /// </summary>
-        protected PlaceholderState PlaceholderState
-        {
-            get => placeholderState;
-            set
-            {
-                if (value != PlaceholderState.Successful)
-                {
-                    Reset();
-                }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:147]: Yes, I wanted to remove this nonsense from the placeholder flow. Was really hard to parse.

**Comment by peppy** [comment]
> Difficult diff to go through. I see an improvement in quality if I squint but it feels somewhat marginal if I'm honest - maybe because the diff is that large and not much was really changed, a lot of stuff has just been shuffled around.

I dunno. Try viewing the files not the diff. The previous flow of execution was atrocious. I'd argue against "not much has changed", with particular focus to the cancellation logic and number of Get/Fetch/Update/whatever methods. This class has been a constant source of issues if you blame back, and had no short of four cancellation tokens/delegates prior to this PR.

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 5]
```diff
@@ -231,96 +145,183 @@ protected Leaderboard()
...
+        {
+            currentScoresAsyncLoadCancellationSource?.Cancel();
 
-        private Placeholder currentPlaceholder;
+            scoreFlowContainer?
+                .FadeOut(fade_duration, Easing.OutQuint)
+                .Expire();
+            scoreFlowContainer = null;
+
+            loading.Hide();
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:291]: I believe the one in the placeholder flow can be removed. Will look to do this.

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:317]: It wouldn't be the same thing, would it, though? To keep this exactly as is you'd have to do

```csharp
using (s.BeginDelayedSequence(i * 50)
    s.Show();
i += 1;
```

which shows the problem with the original code (it tries to be a bit too clever for my liking).

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:317]: Sure I'll convert it to plain math. We should probably discuss this at some point since you take issue with this but I prefer the way it's written there. I'll try and stop doing that for this project to keep everyone happy.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: Should probably tag in @smoogipoo for this. Maybe it's just me finding this difficult, in which case I'll bite the bullet and attempt to tolerate these.

**Comment by bdach** [comment]
> Try viewing the files not the diff. The previous flow of execution was atrocious. I'd argue against "not much has changed", with particular focus to the cancellation logic and number of Get/Fetch/Update/whatever methods.

Okay, I'll take that back, I can see your point there, it does look much better upon quickly skimming current `master`. I did definitely want to look over the diff first to check for any behaviour differences though, and that was much obfuscated by member renaming/reordering which is why I said what I did initially.

**Comment by bdach** [comment]
Also I've just re-read the OP and it mentions this:

> Also note that the tests I added somewhat duplicate those in `TestSceneBeatmapLeaderboard`, but I intentionally wanted to cover the full flow, including affirming that the local play is recorded to the database, shows up at song select on returning, and can be interacted with. Nice high level coverage to have.

Which ones would those be? I can only see small changes in `TestSceneBeatmapLeaderboard` and `TestSceneDeleteLocalScore`, did something not get pushed out or got stuck in a stash?

**Comment by peppy** [comment]
Actually that comment is from my notes which was supposed to be in #16681 - the mentioned tests were merged as part of that already 😅 .

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:293]: Okay so 3d771c0f removes a *second* unconditional hide, so I can see how this is relevant in a degree, but my original gripe was a little different.

When `Scores.set` calls `updateScoresDrawables()` with actual scores, two hides happen inside the latter method:

https://github.com/ppy/osu/blob/d3cb910cf82cb208bf3ae6c3059e8b851107770e/osu.Game/Online/Leaderboards/Leaderboard.cs#L293

https://github.com/ppy/osu/blob/d3cb910cf82cb208bf3ae6c3059e8b851107770e/osu.Game/Online/Leaderboards/Leaderboard.cs#L304-L326

Note that the second one is near the end of the async load completion callback. If the async load of new content takes long enough, then there will be a period where the spinner is hidden but the new content wasn't there yet. This isn't an issue on `master` as `master` does this:

https://github.com/ppy/osu/blob/4132f67629467f5a5f2b2b480b391525bfe7d95e/osu.Game/Online/Leaderboards/Leaderboard.cs#L64-L68

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:306]: Fair call. I'll move it inside the early return conditional.

**Comment by peppy** [comment]
I'm going to do a second pass on the placeholder logic now that I've started to fix things there. Probably easier to bundle in this PR given how pointless the diff is already.

**Comment by peppy** [comment]
Have completed an initial cleanup pass on the "placeholder" logic flow.

**Comment by peppy** [comment]
Note that I force pushed just the last commit, in order to fix this PR being in a bugged state due to github woes. Should not affect review history as it wasn't displayed from the outset.

**Comment by bdach** [comment]
Loading spinner logic seems to have broken when there is no user logged in:

2022-01-30-142112_552x137_scrot

Shouldn't take anything special to reproduce, should be sufficient to log out and select an online leaderboard tab on song select.

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 6]
```diff
@@ -231,120 +128,224 @@ protected Leaderboard()
...
+            Debug.Assert(scores?.Any() != true);
 
-        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+            setErrorState(errorState);
+        }
 
-        [BackgroundDependencyLoader]
-        private void load()
+        /// <summary>
+        /// Call when score retrieval is ready to be displayed.
```
Comment (bdach) [inline osu.Game/Online/Leaderboards/Leaderboard.cs:179]: kind of reads weird

```suggestion
        /// Call when retrieved scores are ready to be displayed.
```

File: osu.Game/Online/Leaderboards/LeaderboardErrorState.cs [hunk 7]
```diff
@@ -3,10 +3,9 @@
 
 namespace osu.Game.Online.Leaderboards
 {
-    public enum PlaceholderState
+    public enum LeaderboardErrorState
     {
-        Successful,
-        Retrieving,
```
Comment (bdach) [inline osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: not super sure about removing `Retrieving` here - you could fold all of the loading spinner logic into the leaderboard state management rather easily, i think (on every state change, show the spinner if state is `Retrieving`, otherwise hide it - would also fix the bug mentioned above).

**Review by bdach — COMMENTED** [review]
latest changes look generally sane / an improvement

File: osu.Game/Online/Leaderboards/LeaderboardErrorState.cs (hunk 7 above)
Comment (peppy) [reply osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: Main reason for removing it was scoping the naming of this enum. I really didn't like the previous name, and also wanted to make it exclusively for error states if possible.

Comment (bdach) [reply osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: I think that in this particular case there's more value in having the enum represent the general state of the entire control (including the intermediate loading state) than just the end result of a score fetch, but see how you go in fixing the loading spinner visibility thing I guess. Maybe it can be done cleanly without - the main problem I see there with how things are structured right now is that `NoError` is going to be ambiguous (because it can either mean "scores have been fetched OK", or "scores are being fetched right now").

Comment (peppy) [reply osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: What would you call the enum if `Retrieving` was added back?

Comment (bdach) [reply osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: The entire thing you mean? Probably something like `LeaderboardDisplayState` or just even `LeaderboardState`.

Comment (peppy) [reply osu.Game/Online/Leaderboards/LeaderboardErrorState.cs:9]: I'll give it a try.

File: osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs [hunk 8]
```diff
@@ -241,10 +164,53 @@ protected override LeaderboardScore CreateDrawableTopScore(ScoreInfo model) => n
...
+                    // we're creating and using a string list representation of selected mods so that it can be translated into the DB query itself
+                    var selectedMods = mods.Value.Select(m => m.Acronym);
+                    scores = scores.Where(s => s.Mods.Any(m => selectedMods.Contains(m.Acronym)));
+                }
+
+                scores = scores.Detach();
+
+                scoreManager.OrderByTotalScoreAsync(scores.ToArray(), cancellationToken)
+                            .ContinueWith(ordered =>
+                            {
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs:205]: I believe this needs a cancellation check? The callback is still fired asynchronously.

File: osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs [hunk 9]
```diff
@@ -241,10 +164,53 @@ protected override LeaderboardScore CreateDrawableTopScore(ScoreInfo model) => n
...
+                    // otherwise find all the scores that have *any* of the currently selected mods (similar to how web applies mod filters)
+                    // we're creating and using a string list representation of selected mods so that it can be translated into the DB query itself
+                    var selectedMods = mods.Value.Select(m => m.Acronym);
+                    scores = scores.Where(s => s.Mods.Any(m => selectedMods.Contains(m.Acronym)));
+                }
+
+                scores = scores.Detach();
+
+                scoreManager.OrderByTotalScoreAsync(scores.ToArray(), cancellationToken)
+                            .ContinueWith(ordered =>
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs:204]: This looks like it needs to be scheduled. Reasoning being is that this is async and the implementation of `SetScores` is doing update-bound tasks (`Show()`/`Hide()` at least).

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 10]
```diff
@@ -213,11 +110,11 @@ protected Leaderboard()
                             },
                             new Drawable[]
                             {
-                                content = new Container
+                                new Container
```
Comment (smoogipoo) [inline osu.Game/Online/Leaderboards/Leaderboard.cs:113]: Is this container required anymore?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: Definitely like the way it is currently in this PR (`delay += 50` after iteration) than in the above comment or in the original diff.

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 10 above)
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs:113]: Doesn't look like it.

File: osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs:204]: It's already safe to some level due to the synchronisation context, but will add a schedule to localise execution I guess.

Comment (peppy) [reply osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs:204]: Actually it isn't due to the use of `ContinueWith`, and not `await`+`ConfigureAwait(true)`

**Review by smoogipoo — APPROVED** [review]
